File: Data Structures & Algorithms/decode-ways/submission-0.cpp
class Solution {
    unordered_map<int, int> cache;
public:
    int numDecodings(string s) {
        return numWays(s, 0);
    }
    int numWays(string& s, int l) {
        if(cache.count(l)) {
            return cache[l];
        }

        if(s.size() - l == 0) {
            return 1;
        }
        if(s[l] == '0') {
            return 0;
        }
        int res = -1;
        if(s[l] == '1') {
            res = numWays(s, l + 1) + numWays(s, l + 2);
            cache[l] = res;
            return res;
        }
        if(s[l] == '2') {
            if(s[l + 1] - '0' > 6) {
                res = numWays(s, l + 1);
                cache[l] = res;
                return res;
            } else {
                res = numWays(s, l + 1) + numWays(s, l + 2);
                cache[l] = res;
                return res;
            }
        }
        res = numWays(s, l + 1);
        cache[l] = res;
        return res;
    }
};

File: Data Structures & Algorithms/spiral-matrix/submission-0.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        
        m, n = len(matrix), len(matrix[0])
        u, d, l, r = 0, m - 1, 0, n - 1

        res = []

        while u <= d and l <= r:

            for i in range(l, r + 1):
                res.append(matrix[u][i])

            u += 1

            for i in range(u, d + 1):
                res.append(matrix[i][r])

            r -= 1

            if u > d or l > r:
                break

            for i in range(r, l - 1, -1):
                res.append(matrix[d][i])

            d -= 1

            for i in range(d, u - 1, -1):
                res.append(matrix[i][l])

            l += 1

        return res